packages/vm/src/context.rs

```diff
@@ -135,13 +163,38 @@ pub(crate) fn move_into_context<S: Storage, Q: Querier>(target: &mut Ctx, storag
     b.querier = Some(querier);
 }
 
+pub fn get_gas_status<S: Storage, Q: Querier>(ctx: &mut Ctx) -> &mut GasState {
+    &mut get_context_data::<S, Q>(ctx).gas_state
+}
+
+pub fn try_consume_gas<S: Storage, Q: Querier>(ctx: &mut Ctx, used_gas: u64) -> VmResult<()> {
+    use crate::backends::get_gas_left;
+
+    let ctx_data = get_context_data::<S, Q>(ctx);
+    if let Some(mut instance_ptr) = ctx_data.wasmer_instance {
+        let instance = unsafe { instance_ptr.as_mut() };
+
+        let gas_status = get_gas_status::<S, Q>(ctx);
+        let wasmer_used_gas = gas_status.gas_limit - get_gas_left(instance);
+
+        gas_status.use_gas(used_gas);
```

Comment:
shouldn't the `instance`'s gas left be updated as well? Otherwise the externally used gas can be re-used as part of the following wasm computation

Reply:
can you elaborate?
I'm not updating the instance gas because i trigger `GasDepletion` when i see that the sum of gas used inside the contract and outside the contract exceed the total gas limit. I didn't see a need to update it at all

Reply:
We set a gas limit to the Wasmer instance to ensure the execution stops when out of gas. Now you realized correctly that calls out if the contract are free, which should not be the case. Calling `try_consume_gas` correctly terminates the execution when the gas limit is exceeded. But if gas is consumed and there is gas left, we need a way to report the gas consumptin to Wasmer. Having the value as a context value is not sufficient. Otherwise you can spend the gas twice: in a call to the backend and as computational resources after that.

Reply:
Yeah, my question is what could that gs be used for after?
If you're really concerned about this, we can set the gas limit in wasmer to `0` after observing a contract-halt due to `GasDepletion`

Reply:
I'm not talking about the `Err` case but the `Ok` case

Reply:
Oh, you mean after the last call to the external services, and before finishing the computation? that's a good point...

Reply:
yeah basically every `try_consume_gas` would need to be synced back to Wasmer since you never know if Wasmer of the external call will be the end of gas

Reply:
I guess this explains the remaining test fail as well:

```
failures:

---- singlepass_tests::handle_storage_loop stdout ----
thread 'singlepass_tests::handle_storage_loop' panicked at 'assertion failed: `(left == right)`
  left: `2054`,
 right: `0`', tests/integration.rs:317:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The storage call cost are not reported to the Wasmer instance, which is the source of `Instance::get_gas_left`

Reply:
I did not notice this failure before. Let me fix one thing at a time - if that's still there after, then i'll address it as well :sweat_smile:

Reply:
Yeah, looks like this error was also because of the errors from my 2AM coding session :smile: 
fixed in 9c2a00a.